File: functional_concepts/composition/composition.go
package main

import "fmt"

type any interface{}
type function func(any) any

func compose(f, g function) function {
	return func(value any) any {
		return f(g(value))
	}
}

func square(x any) any {
	return x.(int) * x.(int)
}

func main() {
	fmt.Printf("compose(square, square)(2) = %d\n", compose(square, square)(2))
	fmt.Printf("compose(square, square)(0) = %d", compose(square, square)(0))
}

File: functional_concepts/composition/composition.ex
compose = fn f, g ->
  fn value -> f.(g.(value)) end
end

square = fn x -> x * x end

IO.puts "compose(square, square)(2) = #{compose.(square, square).(2)}" # == 16
IO.puts "compose(square, square)(1) = #{compose.(square, square).(1)}" # == 1
